feat: Implement --local-git-group-override with group match on repo path

epcim opened the pull request:
# Description

This is early draft,

@codablock not sure I did fully your hostname/path comparison. There are multiple variants that can be implemented. I would kindly ask for new round of comments - as I have bit feeling it was drifting wrong way.

you shall either clarify, what we want:
- only to replace matching git url to local path
- even allow any part of the git url to be replaced by any override (ie github.com to gitlab.com/xyz etc..)

## New features
- git local group override, git url pattern to override external url to local dir
- ... TBD

## Fixes
- fixes initialisation of overriden repos

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [ ] This change requires a new example

<!---
All Submissions:

* [ ] A corresponding issue exists
* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open Pull Requests for the same update/change?
* [ ] I have performed a self-review of my code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] Have you lint your code locally before submission?
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules
* [ ] All commits are signed off which certify that you created the patch and that you agree to the [Developer Certificate of Origin](https://developercertificate.org/)
-->

epcim commented on cmd/kluctl/args/project.go:
```diff
@@ -27,7 +27,8 @@ type ProjectFlags struct {
 	GitCacheUpdateInterval time.Duration `group:"project" help:"Specify the time to wait between git cache updates. Defaults to not wait at all and always updating caches."`
-	LocalGitOverride       []string      `group:"project" help:"Specify a local git override in the form of 'github.com:my-org/my-repo=/local/path/to/override'. This will cause kluctl to not use git to clone for the specified repository but instead use the local directory. This is useful in case you need to test out changes in external git repositories without pushing them. To only override a single ...
+	LocalGitOverride       []string      `group:"project" help:"Specify a single repository local git override in the form of 'github.com:my-org/my-repo=/local/path/to/override'. This will cause kluctl to not use git to clone for the specified repository but instead use the local directory. This is useful in case you need to test out changes in external git repositories without pushing them. To only ...
+	LocalGitGroupOverride  []string      `group:"project" help:"Specify a pattern for git repository url override'. Usage ase --local-git-override. Affected are all repositories matching the pattern.`
```
needs update, once use-case description is improved

epcim commented on cmd/kluctl/commands/utils.go:
```diff
@@ -246,6 +254,30 @@ func parseRepoOverride(s string) (ret repocache.RepoOverride, err error) {
 	return
 }
+
+func parseRepoGroupOverride(s string) (ret repocache.RepoOverride, err error) {
```
copy/paste :/ 
could reuse original parseRepoGroup() but the erorr msg would need update. 

for now my intention is to keep it separate, as if use-case would be more then 1:1 replacement it might need different code?

codablock replied:
Regarding the error message, you can completely remove the `--local-git-override` from the returned error and instead let `withKluctlProjectFromArgs` wrap the returned error messages with the different flag names, e.g.:
```go
ro, err := parseRepoOverride(x, true) // true for --local-git-group-override
if err != nil {
	return fmt.Errorf("invalid --local-git-group-override: %w", err)
}
```

epcim commented on pkg/helm/helm_release.go:
```diff
@@ -238,6 +240,17 @@ func (hr *Release) doRender(ctx context.Context, k *k8s.K8sCluster, k8sVersion s
 		return err
 	}
 
+	// FIXME, alternative path under helm repo
```
ignore this

epcim commented on pkg/repocache/cache.go:
```diff
@@ -85,7 +88,48 @@ func (rp *GitRepoCache) GetEntry(url git_url.GitUrl) (*CacheEntry, error) {
+		}
+		uN := uO.NormalizedRepoKey()
+
+		// FIXME, what was the requested logic here?
```
can you update the requested logic herE?

codablock replied:
I'll try to checkout your branch and add a commit that you can then cherry-pick.

codablock commented on cmd/kluctl/commands/utils.go:
```diff
@@ -60,6 +60,14 @@ func withKluctlProjectFromArgs(ctx context.Context, projectFlags args.ProjectFla
+		if err != nil {
+			return err
+		}
+		ro.IsGroup = true
```
Better to put this into `parseRepoGroupOverride`

epcim replied:
ok, you know I had opposite idea :D

codablock commented on cmd/kluctl/commands/utils.go:
```diff
@@ -246,6 +254,30 @@ func parseRepoOverride(s string) (ret repocache.RepoOverride, err error) {
 		ret.Ref = sp2[2]
 	}
-	ret.Override = sp[1]
+	ret.Override = 	os.ExpandEnv(sp[1])
```
Kluctl should better not do `ExpandEnv` and instead rely on the calling shell to do this properly.

epcim replied:
make sense if used on CLI, less if it would come as part of the config in .kluctl.yml once.
but agree - for now will revert

codablock replied:
I'm not sure we should have such a feature as part of the .kluctl.yml, as that would not be in the scope of what .kluctl.yml tries to achieve (defining targets).

If you want to have such options as part of your source/repo, you can always revert to tools like https://github.com/direnv/direnv which would then include the necessary environment variables to be active while you're inside the project dir. That's how we solve such things usually.

codablock commented on pkg/types/git_project.go:
```diff
@@ -24,8 +25,10 @@ func (gp *GitProject) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	gp := sl.Current().Interface().(GitProject)
-	if strings.Index(gp.SubDir, "/") != -1 {
-		sl.ReportError(gp.SubDir, "subDir", "SubDir", "/ is not allowed in subDir", "")
+	for _, e := range filepath.SplitList(gp.SubDir) {
```
`filepath.SplitList` splits by `:` and `;`. What you want to do is a plain `strings.Split(gp.SubDir, "/")`

codablock commented:
@epcim I pushed a few commits to the branch https://github.com/kluctl/kluctl/tree/improveLocal-git-override which you can cherry-pick.

Regarding the question if overrides to other repo urls should be possible: In general I would say yes, that might be a good feature as well, but I would prefer to have this in another PR after these changes are finished.

codablock commented on pkg/types/git_project.go:
```diff
@@ -24,8 +26,10 @@ func (gp *GitProject) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	gp := sl.Current().Interface().(GitProject)
-	if strings.Index(gp.SubDir, "/") != -1 {
-		sl.ReportError(gp.SubDir, "subDir", "SubDir", "/ is not allowed in subDir", "")
+	for _, e := range strings.Split(gp.SubDir, string(os.PathSeparator)) {
```
The subPath in question here is not os-dependent so it must split by `/`.
While writing this I realised that this validation should also check for `\` to not be existent in the path and that there might also be many other places where such checks would be required.

This sounds like it should be moved into a completely new PR with only path related fixes. Do you want me to take over that part or do you also want to do that? I think I will find some time tomorrow to work on Kluctl topics again.

epcim replied:
I can have a look. Then we probably want to match it some regex pattern. `[-_\.a-zA-Z0-9/]+` will have to check whether at least part is fully covered by `\w` in golang. Though regex will take more cpu cycles however on dedicated fuction it could be reasonable.

codablock reviewed (commented):
Overall LGTM, just the single comment I added should be handled.

The new flag would need some e2e tests as well. If you're still motivated, you can add those to this PR (I will assist you of course). As an alternative we can merge this PR and I will later add e2e tests.

codablock commented:
fyi, you'll need to run `make replace-commands-help` and commit the changes to make the tests succeed.

epcim commented:
> Overall LGTM, just the single comment I added should be handled.
> 
> The new flag would need some e2e tests as well. If you're still motivated, you can add those to this PR (I will assist you of course). As an alternative we can merge this PR and I will later add e2e tests.

well I still count to do some more testing on my side, that's why I kept it as draft

e2e I checked, but as there were not git-local-override existing tests I did not wrote any. Any existing test there I could re-configure and use or it needs new test-case? Most probably I would keep this for separate PR (even I dont like that approach).

codablock commented:
No there is no existing test that could be used for this, so a new one would be needed. I assume that some additional work will be required for the e2e tests as the test code is currently not able to work with multiple repos. I'll look into that after this PR is merged.

Please ping me when you feel this PR is ready for another round of review.

epcim commented:
> No there is no existing test that could be used for this, so a new one would be needed. I assume that some additional work will be required for the e2e tests as the test code is currently not able to work with multiple repos. I'll look into that after this PR is merged.
> 
> Please ping me when you feel this PR is ready for another round of review.

Works like a charm on my side for the use-case I have. 

```
export KLUCTL_LOCAL_GIT_GROUP_OVERRIDE_0=gitlab.com:myorg/mysub=$PWD/.repos
```

```
❮ kluctl deploy -t "crt" --dry-run
✓ Loading kluctl project
✓ Initializing k8s client
⚠ Overriding git repo ssh://git@gitlab.com/myorg/mysub/lma-deployment with local directory /Users/xxx/Workspace/gitlab/myorg/mysub/sre-crt-model/deployment/kube/.repos/lma-deployment
⚠ Overriding git repo ssh://git@gitlab.com/myorg/mysub/cicd-deployment with local directory /Users/xxx/Workspace/gitlab/myorg/mysub/sre-crt-model/deployment/kube/.repos/cicd-deployment
✓ Rendering templates
✓ Rendering Helm Charts
✓ Building kustomize objects
✓ Postprocessing objects
✓ Getting 5 additional remote objects
✓ Getting namespaces
✓ secrets: Nothing to apply.
✓ justTest: Applied 1 objects.
✓ ves-system: Applied 1 objects.
✓ falcosidekick: Applied 3 objects.
✓ Running server-side apply for all objects
...
...

```

epcim commented:
> @epcim Very cool, happy to see that your use case is served well :) Also very happy to see contributors, so big thanks to you 👍
> 
> One little thing that would need to be changed: ValidateGitProject should be completely reverted as it now accidentally got the erroneous SplitList code.
> 
> After that, I will squash-merge the MR. I will also change the title to include "feat: " as a semantic version prefix.

though I already reverted somewhere on one of my forks, reverted now.
sure squash/change as needed.

my pleasure to get hands on this - this project addresses all my thoughts/needs from last two years.

codablock commented on cmd/kluctl/args/project.go:
```diff
@@ -27,7 +27,8 @@ type ProjectFlags struct {
 	GitCacheUpdateInterval time.Duration `group:"project" help:"Specify the time to wait between git cache updates. Defaults to not wait at all and always updating caches."`
-	LocalGitOverride       []string      `group:"project" help:"Specify a local git override in the form of 'github.com:my-org/my-repo=/local/path/to/override'. This will cause kluctl to not use git to clone for the specified repository but instead use the local directory. This is useful in case you need to test out changes in external git repositories without pushing them. To only override a single ...
+	LocalGitOverride       []string      `group:"project" help:"Specify a single repository local git override in the form of 'github.com:my-org/my-repo=/local/path/to/override'. This will cause kluctl to not use git to clone for the specified repository but instead use the local directory. This is useful in case you need to test out changes in external git repositories without pushing them. To only ...
+	LocalGitGroupOverride  []string      `group:"project" help:"Specify a pattern for any git repository url override'. Usage 'gitlab.com:others/sub-org=/local/path/to/my-forks'. Affected are all source repositories which url/path is matching "gitlab.com:others/sub-org" pattern.`
```
Can you change the description to:
```
Same as --local-git-override, but for a whole group prefix instead of a single repository. All repositories that have the given prefix will be overridden with the given local path and the repository suffix appended. For example, 'gitlab.com:some-org/sub-org=/local/path/to/my-forks' will override all repositories below 'gitlab.com:some-org/sub-org/' with the repositories found in '/local/path/to/my-forks'. It will however only perform an override if the given repository actually exists locally and otherwise revert to the actual (non-overridden) repository.
```

Don't forget to call `make replace-commands-help` afterwards.

epcim replied:
done, agree - better wording and example

codablock commented:
One last nit I found while checking the description, after that I'll merge the PR :)
